### PR TITLE
Endpoint user management for products & repositories

### PR DESCRIPTION
**Pull request by Juli0q**

### CODE REVIEW

File: core/src/test/kotlin/api/ProductsRouteIntegrationTest.kt [hunk 1]
```diff
@@ -676,4 +695,231 @@ class ProductsRouteIntegrationTest : AbstractIntegrationTest({
...
+        forAll(
+            row(HttpMethod.Put),
+            row(HttpMethod.Delete)
+        ) { method ->
+            "require ProductPermission.WRITE for method '${method.value}'" {
+                val createdProd = createProduct()
+                val user = Username(TEST_USER.username.value)
+                print("user: $user")
+                print("createdOrg: $createdProd")
+                print("createdOrg.id: ${createdProd.id}")
```
Comment (sschuberth) [inline core/src/test/kotlin/api/ProductsRouteIntegrationTest.kt:709]: We should not do console output like this in tests. Instead, consider regular logging, or [withClue()](https://kotest.io/docs/assertions/clues.html) is only required in the context of a failed assertion.

Comment (Juli0q) [reply core/src/test/kotlin/api/ProductsRouteIntegrationTest.kt]: That was debug output I forgot to remove, I've removed it now 🙂

File: core/src/test/kotlin/api/ProductsRouteIntegrationTest.kt [hunk 2]
```diff
@@ -676,4 +695,228 @@ class ProductsRouteIntegrationTest : AbstractIntegrationTest({
...
+                        ) {
+                            setBody(user)
+                        }
+                        else -> error("Unsupported method: $method")
+                    }
+
+                    response shouldHaveStatus HttpStatusCode.NotFound
+
+                    val body = response.body<ErrorResponse>()
+                    println(body)
```
Comment (sschuberth) [inline core/src/test/kotlin/api/ProductsRouteIntegrationTest.kt:779]: One more `println()` call to remove.

File: core/src/test/kotlin/api/RepositoriesRouteIntegrationTest.kt [hunk 3]
```diff
@@ -876,5 +893,229 @@ class RepositoriesRouteIntegrationTest : AbstractIntegrationTest({
...
+                            ) {
+                                setBody(user)
+                            }
+                            else -> error("Unsupported method: $method")
+                        }
+
+                        response shouldHaveStatus HttpStatusCode.NotFound
+
+                        val body = response.body<ErrorResponse>()
+                        println(body)
```
Comment (sschuberth) [inline core/src/test/kotlin/api/RepositoriesRouteIntegrationTest.kt:977]: One more `println()` call to remove.

File: core/src/main/kotlin/api/OrganizationsRoute.kt
Comment (mnonnenmacher) [inline core/src/main/kotlin/api/OrganizationsRoute.kt:1]: According to the [contributing guide](https://github.com/eclipse-apoapsis/ort-server/blob/main/CONTRIBUTING.md#atomic-commits) the independent refactorings in this commit should be in separate commits:

* Rename the endpoints in the API docs
* Remove the unnecessary println from the test
* Remove the unnecessary organizationId argument from the functions in the AuthorizationService
* Move the `toJoinedString` helper function

File: core/src/main/kotlin/apiDocs/ProductsDocs.kt [hunk 4]
```diff
@@ -174,7 +174,7 @@ val postRepository: OpenApiRoute.() -> Unit = {
         jsonBody<CreateRepository> {
             example("Create repository") {
                 value = CreateRepository(
-                    type = RepositoryType.GIT,
+                    type = GIT,
```
Comment (mnonnenmacher) [inline core/src/main/kotlin/apiDocs/ProductsDocs.kt:177]: Unrelated change.

File: core/src/main/kotlin/apiDocs/ProductsDocs.kt [hunk 5]
```diff
@@ -190,7 +190,7 @@ val postRepository: OpenApiRoute.() -> Unit = {
                         id = 1,
                         organizationId = 2,
                         productId = 3,
-                        type = RepositoryType.GIT,
+                        type = GIT,
```
Comment (mnonnenmacher) [inline core/src/main/kotlin/apiDocs/ProductsDocs.kt:193]: Unrelated change.

File: core/src/main/kotlin/apiDocs/ProductsDocs.kt [hunk 6]
```diff
@@ -348,3 +348,65 @@ val deleteSecretByProductIdAndName: OpenApiRoute.() -> Unit = {
...
+        jsonBody<Username> {
+            example("Remove user identified by username 'abc123'.") {
+                value = Username(username = "abc123")
+            }
+        }
+    }
+
+    response {
+        HttpStatusCode.NoContent to {
+            description = "Successfully removed the user to the group."
```
Comment (mnonnenmacher) [inline core/src/main/kotlin/apiDocs/ProductsDocs.kt:405]: "**from** the group"

File: services/hierarchy/src/main/kotlin/ProductService.kt [hunk 7]
```diff
@@ -100,4 +102,60 @@ class ProductService(
...
+            )
+        }
+
+        val groupName = productRole.groupName(productId)
+
+        // As the AuthorizationService does not distinguish between technical exceptions (e.g. cannot connect to
+        // Keycloak) and business exceptions (e.g. user not found), we can't do special exception handling here
+        // and just let the exception propagate.
+        authorizationService.removeUserFromGroup(username, groupName)
+    }
```
Comment (mnonnenmacher) [inline services/hierarchy/src/main/kotlin/ProductService.kt:160]: There should also be tests for the new functions, similar to those in `OrganizationServiceTest`.

File: core/src/main/kotlin/api/ProductsRoute.kt [hunk 8]
```diff
@@ -23,6 +23,7 @@ import io.github.smiley4.ktorswaggerui.dsl.routing.delete
 import io.github.smiley4.ktorswaggerui.dsl.routing.get
 import io.github.smiley4.ktorswaggerui.dsl.routing.patch
 import io.github.smiley4.ktorswaggerui.dsl.routing.post
+import io.github.smiley4.ktorswaggerui.dsl.routing.put
```
Comment (mnonnenmacher) [inline core/src/main/kotlin/api/ProductsRoute.kt:26]: Commit message nits:
* plural "endpoints" in title, and maybe change to "Add endpoints to manage product groups"
* "2 new" is redundant.

File: core/src/main/kotlin/api/RepositoriesRoute.kt [hunk 9]
```diff
@@ -23,6 +23,7 @@ import io.github.smiley4.ktorswaggerui.dsl.routing.delete
 import io.github.smiley4.ktorswaggerui.dsl.routing.get
 import io.github.smiley4.ktorswaggerui.dsl.routing.patch
 import io.github.smiley4.ktorswaggerui.dsl.routing.post
+import io.github.smiley4.ktorswaggerui.dsl.routing.put
```
Comment (mnonnenmacher) [inline core/src/main/kotlin/api/RepositoriesRoute.kt:26]: Commit message nits:

* plural "endpoints" in title, and maybe change to "Add endpoints to manage repository groups"
* "2 new" is redundant.

File: core/src/main/kotlin/apiDocs/RepositoriesDocs.kt [hunk 10]
```diff
@@ -649,3 +650,65 @@ val deleteSecretByRepositoryIdAndName: OpenApiRoute.() -> Unit = {
...
+        jsonBody<Username> {
+            example("Remove user identified by username 'abc123'.") {
+                value = Username(username = "abc123")
+            }
+        }
+    }
+
+    response {
+        HttpStatusCode.NoContent to {
+            description = "Successfully removed the user to the group."
```
Comment (mnonnenmacher) [inline core/src/main/kotlin/apiDocs/RepositoriesDocs.kt:707]: "**from** the group"

File: services/hierarchy/src/main/kotlin/RepositoryService.kt [hunk 11]
```diff
@@ -119,4 +121,60 @@ class RepositoryService(
...
+            )
+        }
+
+        val groupName = repositoryRole.groupName(repositoryId)
+
+        // As the AuthorizationService does not distinguish between technical exceptions (e.g. cannot connect to
+        // Keycloak) and business exceptions (e.g. user not found), we can't do special exception handling here
+        // and just let the exception propagate.
+        authorizationService.addUserToGroup(username, groupName)
+    }
```
Comment (mnonnenmacher) [inline services/hierarchy/src/main/kotlin/RepositoryService.kt:179]: There should be tests for the new functions, similar to those in `OrganizationServiceTest`.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]
Not sure why the CI checks did not start, hopefully it works with your next push.

**Review by mnonnenmacher — APPROVED** [review]
